### PR TITLE
Translate remaining Fortran code

### LINKED ISSUE #603: Proposal: rewrite core codebase in C++
I often get asked why OpenMC is written in Fortran instead of C or C++ or another language. When I started the code, @bforget and I were interested primarily in developing and studying parallel algorithms for MC transport aimed at overcoming some of the traditional limitations on scalability. As such, we were naturally very interested in Fortran co-arrays, which had recently been added to the language in the 2008 standard and offered the promise of expressing parallelism through RDMA. I started playing around with co-arrays during the summer of 2011 and quickly discovered that there were a number of major shortcomings:
- There was really only one implementation at the time (Intel), and it wasn't just standard in their compiler; you also had to buy their cluster software (can't remember the name). That would be pretty much in diametric opposition to the idea of OpenMC being free and open if you have to buy software just to run in parallel.
- In F2008, co-array functionality lacks collective operations which is a major problem for scalability (supposedly will be fixed in F2015).

That was pretty much the end of our experiments with co-arrays. However, we had a considerable amount of functionality in OpenMC already built on Fortran, and I didn't want to "waste time" as a PhD student just rewriting what I already had. Anyone who has done a thesis knows that expediency often wins.

Fast forward to present day. OpenMC is still in Fortran and has grown from about 14k lines of code (version 0.4.0) to now 40k lines of code. That may sound like a lot, but by most standards 40k lines of code is actually quite small. MCNP 6.1 is something like ~300k, Serpent2 is ~240k, and Geant4 is something like 1M. I believe there are a number of compelling reasons that we should now consider rewriting the core codebase in C++ while the it is still relatively small:
- **Use of external libraries**: Right now if you want to use a library from OpenMC, you better hope that it has a Fortran API. If not, you better hope that whatever you want to do can be done in a C interoperable manner. Anyone who hasn't been living under a rock knows that most libraries are written in C or C++. Let's take, for example, parallelism, which has been a primary driver in the development of OpenMC. Some of the available options for parallelism in C++ include standard-library features (threads, mutexes, locks, futures, etc.), C++ AMP, Intel TBB, Cilk Plus, Vc library, etc., and that's on top of the usual MPI and OpenMP C interfaces. For Fortran, you really only have MPI and OpenMP. This is not to say that we should be using any of these fancy options, but just merely to point out that we don't have the option even if we wanted to, especially for R&D purposes.
- **Rich standard-library**: Ever looked at dict_header.F90, list_header.F90, or stl_vector.F90 in OpenMC? All are naive attempts to make up for what the Fortran language is missing -- a standard library with containers. Not to mention that string handling in Fortran is unsatisfying.
- **Ability to get closer to hardware**: Since OpenMC is primarily a code intended for advanced R&D, a rewrite in C or C++ would allow researchers to study ideas that require leveraging hardware-specific features, which almost certainly means using some C API or otherwise calling a C function. Again, you might be able to do this with Fortran and the C interoperability module, but it may be painful/limited.
- **Interoperating with other community codes**: There is already interest in collaboration with the INL/[MOOSE](https://github.com/idaholab/moose), ORNL/Shift, and FermiLab/[Geant](http://geant.cern.ch/) teams. All of these projects are C++-based, so there may be benefits to operating in the same language. [PyNE](https://github.com/pyne/pyne), an open-source NE toolkit, also has a substantial amount of C++ code.
- **Moving towards libopenmc**: The development team has had a number of discussions recently about moving OpenMC in the direction of a library that could be called by other software. Again, you could do this in Fortran, but it is a lot more error-prone.
- **Compiler issues**: Now that we are more heavily utilizing F2003 and F2008 issues, we are often running up against compiler bugs and limitations. I've probably reported somewhere around 20 bugs between the gfortran, Intel, PGI, and IBM compilers. A rewrite in C++ won't necessarily mean all problems are gone, especially because I would advocate that we adopt C++11, but some of the more esoteric issues specific to Fortran could be avoided.
- **Talent pool available**: There are obviously many more C++ developers than Fortran developers, and, depending on who you ask, C++ is usually among the top 5 most popular programming languages. Add to this the fact that Fortran, justly or unjustly, conjures negative feelings for many people.

In terms of the mechanics of a rewrite, some code could immediately be removed (all those containers I mentioned, maybe ~3k lines of code), a sizable chunk could easily be rewritten (recent additions to the code that are already OO-style), and then another sizable chunk will require deeper thought. That being said, a few of the modules that fit in that last category are things we are already targeting a refactor for anyway, e.g., geometry and tallies. I don't want to understate the level of effort and time required, but I do think that it might not be as bad as one might naturally think, especially if we're able to get a few people involved.

Some of the risks involved:
- C++ is a large and complex language. There is no doubt about that. When I think about students looking to use OpenMC for their research, they will certainly have to go through a steep learning curve if they haven't used C++ before. But, I think that their time will be better spent learning C++ than learning Fortran and they will end up with a language skill that is beneficial outside of just scientific computing.
- Another language pops up that is way better than C++ and satisfies most of our desires. For example, there are good arguments that Julia, young as it is, is a better language for scientific computing. However, rewriting in Julia would be a lot riskier than C++ in my opinion. The same is likely to be true of other new languages as well.

### LINKED ISSUE #858: pugixml interface doesn't compile with IBM XL Fortran for Blue Gene/Q
I tried to compile the latest version of OpenMC with bgxlf and it fails here:
```
[  6%] Building Fortran object CMakeFiles/pugixml_fortran.dir/src/pugixml/pugixml_f.F90.o
"/home/romano/openmc/src/pugixml/pugixml_f.F90", line 251.54: 1516-169 (S) A variable that is specified as an argument to the intrinsic procedure C_LOC must either have type and type parameters that are interoperable with a C type, or be a nonpolymorphic, scalar variable with no length type parameters.
```
The compiler in this case rejects valid code -- the [variable](https://github.com/mit-crpg/openmc/blob/e9dfad85196aea7b9263efe6ee34d4fd48b81643/src/pugixml/pugixml_f.F90#L242) in question does have an interoperable type, `character(kind=C_CHAR)`. My guess is that it most likely doesn't know what to do with it because it's a deferred-length char. One possible solution is to use the `block` construct as demonstrated [here](http://degenerateconic.com/fortran-c-interoperability-part-2/).

I'm not going to spend time working on a solution for this now since I don't have a present need to run on BG/Q, but thought I'd open an issue in case anyone else wants to run BG/Q.

### LINKED ISSUE #859: Runtime error with PGI community edition 17.10
PGI has a compiler bug that prevents OpenMC from compiling:
```
PGF90-S-0038-Symbol, size, has not been explicitly declared (/home/romano/openmc/src/stl_vector.F90)
  1 inform,   0 warnings,   1 severes, 0 fatal for clear_int
```
I've run into problems with PGI before getting confused when you have local variables defined somewhere that have the same name as an intrinsic (preventing it from being used elsewhere even though it doesn't make any sense). I'm going to file a report with PGI and look into workaround ahead of our 0.9 release.

### PR DESCRIPTION
**Pull request by paulromano**
Well folks, this is it. With this pull request, OpenMC is no longer a Fortran code. After @smharper translated tallies, everything else was straightforward. There's not too much to say about the changes in this PR. The one thing I'll note is that we still have a bit of cleanup to do after the fact (e.g., fixing all our off-by-ones), but I'll start a new issue for that separate from this PR.

I want to thank everyone who was involved in the C++ conversion efforts as this definitely would not have been possible without help! @smharper @nelsonag @shikhar413 @pshriwise @wbinventor

Closes #603.
Closes #858.
Closes #859.

### CODE REVIEW

File: src/tallies/tally.cpp [hunk 1]
```diff
@@ -503,49 +519,374 @@ Tally::init_triggers(pugi::xml_node node, int i_tally)
...
+      s = get_node_value(node_filt, "type", true);
+    }
+
+    // Make sure bins have been set
+    if (s == "energy" || s == "energyout" || s == "mu" || s == "polar"
+      || s == "azimuthal" || s == "mesh" || s == "meshsurface" || s == "universe"
+      || s == "material" || s == "cell" || s == "distribcell" || s == "cellborn"
+      || s == "cellfrom" || s == "surface" || s == "delayedgroup") {
+      if (!check_for_node(node_filt, "bins")) {
+        fatal_error("Bins not set in filter " + std::to_string(filter_id));
```
Comment (smharper) [inline src/tallies/tally.cpp]: We should probably move this check into the `from_xml` methods of each filter.  It'll be a lot of duplication, but if we don't do it I think we're likely to forget to update this check with new filters.

Comment (paulromano) [reply src/tallies/tally.cpp]: Yeah, I was also thinking that as I was rewriting this bit. I'll go ahead and move it.

**Comment by liangjg** [comment]
Just wanted to note that the intel compilation issue reported in #1114 goes away with these commits. And my preliminary tests showed the OpenMP performance issue with Intel becomes insignificant.

| compiler     | openmp threads per process | mpi processes | calculate rate n/s | speedup |
|--------------|----------------------------|---------------|--------------------|---------|
| gcc-7.3.0    | 1                          | 1             | 4236.4             | ref     |
| intel-19.0.2 | 1                          | 1             | 4353.2             | 1.0     |
| gcc-7.3.0    | 1                          | 4             | 15795.2            | 3.7     |
| intel-19.0.2 | 1                          | 4             | 16916.6            | 4.0     |
| gcc-7.3.0    | 1                          | 8             | 30945.3            | 7.3     |
| intel-19.0.2 | 1                          | 8             | 32804.2            | 7.7     |
| gcc-7.3.0    | 1                          | 16            | 41928.5            | 9.9     |
| intel-19.0.2 | 1                          | 16            | 41162.0            | 9.7     |
| gcc-7.3.0    | 4                          | 1             | 17984.9            | 4.2     |
| intel-19.0.2 | 4                          | 1             | 10799.9            | 2.5     |
| gcc-7.3.0    | 8                          | 1             | 36629.1            | 8.6     |
| intel-19.0.2 | 8                          | 1             | 19360.1            | 4.6     |
| gcc-7.3.0    | 16                         | 1             | 56870.0            | 13.4    |
| intel-19.0.2 | 16                         | 1             | 39919.4            | 9.4     |

File: src/tallies/tally.cpp (hunk 1 above)
Comment (paulromano) [reply src/tallies/tally.cpp]: On second look, `get_node_value` (which is called in `from_xml` for each filter) already has a check for whether the node is missing and will call `fatal_error`, so this check is somewhat redundant. How do you feel about just removing it?

Comment (smharper) [reply src/tallies/tally.cpp]: Good point.  Yeah, we can just remove this check.

**Comment by paulromano** [comment]
@liangjg Good to know the major slowdown with Intel goes away. However, the scaling still seems pretty poor compared to gcc, especially at four threads (2.5 vs 4.2). If you ask me, seems like it's still worth looking into further.

**Comment by makeclean** [comment]
@liangjg is there any chance you can share your model? I recently did some scaling studies on a Marvell X2 thunder system it would be handy to maybe try the same file?

**Comment by liangjg** [comment]
@makeclean For your reference, I used a typical PWR pincell model: [openmc/examples/xml/pincell](https://github.com/openmc-dev/openmc/tree/develop/examples/xml/pincell).
@paulromano yeah, the multi-threading calculations seems not very satisfactory. I can do a profiling to explore the potential reasons later.

File: examples/xml/basic/tallies.xml [hunk 2]
```diff
@@ -6,7 +6,6 @@
   </filter>
 
   <filter id="2" type="energy">
-    <bins>0 20.0e6</bins>
```
Comment (smharper) [inline examples/xml/basic/tallies.xml]: Looks like this change accidentally slipped in

**Comment by wbinventor** [comment]
This is amazing! Well done ushering in a brave new world, folks!

File: examples/xml/basic/tallies.xml (hunk 2 above)
Comment (paulromano) [reply examples/xml/basic/tallies.xml]: Oops, just fixed that!

**Comment by paulromano** [comment]
Thanks all for the kind words of support! Certainly if anyone *dis*approved of this change, now would be a bad time to chime in :laughing:

**Comment by makeclean** [comment]
Shame you havent got a big red button on your desk you can hit for when you merge this

**Comment by shikhar413** [comment]
:beer::beer::beer: So nice not to see any mention of Fortran when building OpenMC! 

I went ahead and tried testing some larger BEAVRS cases and the results don't change between OpenMC++ and OpenMC, can't wait to start using this version for my research :)

**Comment by pshriwise** [comment]
Is there a neon "Now Fortran-free!" sticker we can put on the website in the next release?

**Comment by smharper** [comment]
I don't want to call it too early... but I think we may have a consensus here.  Merging!

**Pull request merged**